sdks/java/core/src/main/java/org/apache/beam/sdk/testing/TestPipeline.java

```diff
@@ -239,8 +246,53 @@ private TestPipeline(
   public Statement apply(final Statement statement, final Description description) {
     return new Statement() {
 
+      private final Predicate<Annotation> hasCategoryAnnotation = new Predicate<Annotation>() {
+
+        @Override
+        public boolean apply(final Annotation annotation) {
+          return annotation.annotationType() != null
+              && annotation.annotationType().equals(Category.class);
+        }
+      };
+
+      private final Predicate<Annotation> categoryOfNeedsRunner = new Predicate<Annotation>() {
+
+        @Override
+        public boolean apply(final Annotation category) {
+          return
+              FluentIterable
+                  .from(Arrays.asList(((Category) category).value()))
+                  .anyMatch(new Predicate<Class<?>>() {
+
+                    @Override
+                    public boolean apply(final Class<?> aClass) {
+                      return NeedsRunner.class.isAssignableFrom(aClass);
+                    }
+                  });
+        }
+      };
+
+      private void setDeducedEnforcementLevel() {
+        if (enforcement instanceof PipelineAbandonedNodeEnforcement) {
```

Comment:
I don't think I understand your first comment - if a user writes `@Rule TestPipeline = TestPipeline.create().enableAbandonedNodeEnforcement(...)` it will still be automatically downgraded if `crashingRunner && !annotatedWithNeedsRunner`. This probably means that a user forgot `NeedsRunner` or shouldn't be using `TestPipeline`, so should fail instead.

Regardless of how we isolate the use cases, I think the whole thing will probably be more readable if it is just very explicit whether it has been set versus inferred. I tend to prefer `Optional` to `@Nullable` for safety, readability and composability.